[network/ebpf] https enable arm64 on kernel >= 5.5.0

nplanel opened the pull request:
### What does this PR do?

https enable arm64 on kernel >= 5.5.0 with runtime compilation enabled

### Motivation

enable https monitoring on arm64

### Describe how to test/QA your changes

```
system_probe_config:
  log_level: debug
  enable_runtime_compiler: true
network_config:
  enable_http_monitoring: true
  enable_https_monitoring: true
```

 * run system-probe on arm64 with a kernel >= 5.5.0 ; like debian11 or ubuntu 21.10 or ami 5.10
 * `wget https://httpbin.org/anything/foo/200`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections | tee cnx | jq '.conns[].tags , .tags'`
You should see openssl tags : `tls.library:openssl`
 * `sudo curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/debug/http_monitoring`


- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

nplanel committed:
enable https probes on arm

nplanel committed:
add (make sure) wget curl packages are installed

nplanel committed:
bpf_probe_read_user() on http/lib.so sys_open >= 5.5.x runtime compilation

nplanel committed:
use bpf_probe_read_user() on kernel >= 5.5.0 and not in prebuild

nplanel committed:
https arm64 support only kernel >= 5.5.0 and runtime compilation enabled

brycekahle commented on test/kitchen/site-cookbooks/dd-system-probe-check/recipes/linux.rb:
```diff
@@ -50,6 +50,10 @@
 
+package 'wget'
+
+package 'curl'
```
Are these related to this PR?

brycekahle commented on pkg/ebpf/c/bpf_helpers.h:
```diff
@@ -84,6 +84,17 @@ static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void
+/* bpf_probe_read_from_user is an helper to use bpf_probe_read_user() on kernel >= 5.5.0
+ * but will use bpf_probe_read() on older and prebuild build.
+ */
+#ifdef DD_PREBUILD
```
I think you could do something simpler, and more transparent like:
```
#if defined(DD_PREBUILD) || LINUX_VERSION_CODE < KERNEL_VERSION(5, 5, 0))
static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read;
static int (*bpf_probe_read_kernel)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read;
#else
static int (*bpf_probe_read_user)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read_user;
static int (*bpf_probe_read_kernel)(void* dst, int size, void* unsafe_ptr) = (void*)BPF_FUNC_probe_read_kernel;
#endif
```
and then use `bpf_probe_read_user` everywhere rather than a new function.

brycekahle replied:
Alternatively, we can try upstreaming a transparent replacement at the loader level in `cilium/ebpf` like what `libbpf` does.

nplanel replied:
cilium/ebpf load will fixup bpf_probe_read_{user,kernel} to bpf_probe_read if kernel doesn't support _{user,kernel}
so the fix will be to rename all our bpf_probe_read(), the loader will do the fixup https://github.com/cilium/ebpf/blob/master/linker.go#L196

brycekahle replied:
Oh nice! I didn't realize that was already there.

brycekahle replied:
Nevermind, they removed the check for `kern_version` in 5.0

nplanel replied:
I seen that libbpf do the same kind of fixup

nplanel committed:
hook SSL_connect() like SSL_do_handshake()

issue found with wget 1.14/openssl 1.0.2k

arm64/amazonlinux2/4.14 kernel running 5.5.0 kernel

nplanel committed:
release note

maycmlee commented on releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml:
```diff
@@ -0,0 +1,12 @@
+---
+enhancements:
+  - |
+    enable https monitoring on arm64 with kernel >= 5.5.0
```
```suggestion
    Enable https monitoring on arm64 with kernel >= 5.5.0,
```

maycmlee commented on releasenotes/notes/trace-agent-enable-https-monitoring-on-arm64-cc3dac41f9c0f28b.yaml:
```diff
@@ -0,0 +1,12 @@
+enhancements:
+  - |
+    enable https monitoring on arm64 with kernel >= 5.5.0
+    runtime compilation is require
```
```suggestion
    runtime compilation is require.
```

brycekahle replied:
```suggestion
    runtime compilation is required.
```

nplanel committed:
s/bpf_probe_read/bpf_probe_read_kernel/g

nplanel committed:
call explicitly bpf_probe_read_{user,kernel} linker will fix the call

nplanel committed:
call bpf_probe_read_user() per byte

If we don't do this arm64 will return EFAULT when out of (userspace) range and will bzero(dst)

nplanel committed:
enable https prebuilt on arm64

nplanel committed:
hook do_sys_openat2 on arm64

brycekahle commented on pkg/network/tracer/tracer_test.go:
```diff
@@ -1628,12 +1646,16 @@ func TestHTTPStats(t *testing.T) {
+	if !httpsSupported(t) {
+		t.Skip("HTTPS feature not available on ARM pre 5.5.0 kernels")
+	}
+	if runningOnARM() && !cfg.EnableRuntimeCompiler {
```
Can't we support prebuilt on ARM (>= 5.5.0) now that the loader does the helper translation for us?

nplanel replied:
yes, the issue is that offset-guessing is not working properly, I will take a look tomorrow

nplanel commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -370,18 +385,15 @@ SEC("kprobe/do_sys_open")
 int kprobe__do_sys_open(struct pt_regs* ctx) {
     char *path_argument = (char *)PT_REGS_PARM2(ctx);
     lib_path_t path = {0};
-    bpf_probe_read(path.buf, sizeof(path.buf), path_argument);
```
bpf_probe_read_user() is basicaly copy_from_user_nofault() and return EFAULT if your outside of userspace boundary. If a EFAULT (page fault) happen the kernel will bzero(dst) :(

brycekahle replied:
Should we be using `bpf_probe_read_user_str` then?

nplanel replied:
`bpf_probe_read_user_str` is supported only from 4.11 and there are no _str fixup/fallback

brycekahle replied:
Is it a matter of retrying in the face of `EFAULT`?

nplanel replied:
I did a fallback as `bpf_probe_read_user` return us an error

nplanel committed:
openat2 prebuilt

nplanel committed:
fill_path_safe() if bpf_probe_read_user() return EFAULT

nplanel committed:
cleanup

nplanel committed:
ebpf/bytecode/runtime hash

brycekahle commented on pkg/network/ebpf/c/runtime/http.c:
```diff
@@ -372,20 +387,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
-        } else if (path.buf[i] == 0) {
-            path.len = i;
+        for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	    if (path.len) {
```
weird indent here

brycekahle commented on pkg/network/ebpf/c/runtime/http.c:
```diff
@@ -372,20 +387,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
+static __always_inline int fill_path_safe(lib_path_t *path, char *path_argument) {
+#pragma unroll
+    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	bpf_probe_read_user(&path->buf[i], 1, &path_argument[i]);
```
weird indent

brycekahle commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -366,20 +381,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
-        } else if (path.buf[i] == 0) {
-            path.len = i;
+        for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	    if (path.len) {
```
weird indent

brycekahle commented on pkg/network/ebpf/c/prebuilt/http.c:
```diff
@@ -366,20 +381,33 @@ int uprobe__gnutls_deinit(struct pt_regs* ctx) {
+static __always_inline int fill_path_safe(lib_path_t *path, char *path_argument) {
+#pragma unroll
+    for (int i = 0; i < LIB_PATH_MAX_SIZE; i++) {
+	bpf_probe_read_user(&path->buf[i], 1, &path_argument[i]);
```
weird indent

brycekahle reviewed (approved):
Some minor formatting issues. Otherwise 👍

nplanel committed:
fixupindent

nplanel committed:
remove bpf_probe_read() compilation trap comment

maycmlee reviewed (approved):
👍 for docs.
